mem/buffer_slice.go

```diff
@@ -0,0 +1,155 @@
+package mem
+
+import (
+	"io"
+)
+
+// BufferSlice offers a means to represent data that spans one or more Buffer
+// instances.
+type BufferSlice []*Buffer
+
+// Len returns the sum of the length of all the Buffers in this slice.
+//
+// # Warning
+//
+// Invoking the built-in len on a BufferSlice will return the number of buffers
+// in the slice, and *not* the value returned by this function.
+func (s BufferSlice) Len() int {
+	var length int
+	for _, b := range s {
+		length += b.Len()
+	}
+	return length
+}
+
+// Ref returns a new BufferSlice containing a new reference of each Buffer in the
+// input slice.
+func (s BufferSlice) Ref() BufferSlice {
+	out := make(BufferSlice, len(s))
+	for i, b := range s {
+		out[i] = b.Ref()
+	}
+	return out
+}
+
+// Free invokes Buffer.Free on each Buffer in the slice.
+func (s BufferSlice) Free() {
+	for _, b := range s {
+		b.Free()
+	}
+}
+
+// CopyTo copies each of the underlying Buffer's data into the given buffer,
+// returning the number of bytes copied.
+func (s BufferSlice) CopyTo(out []byte) int {
+	off := 0
+	for _, b := range s {
+		off += copy(out[off:], b.ReadOnlyData())
+	}
+	return off
+}
+
+// Materialize concatenates all the underlying Buffer's data into a single
+// contiguous buffer using CopyTo.
+func (s BufferSlice) Materialize() []byte {
+	out := make([]byte, s.Len())
+	s.CopyTo(out)
+	return out
+}
+
+// LazyMaterialize functions like Materialize except that it writes the data to a
+// single Buffer pulled from the given BufferPool. As a special case, if the
+// input BufferSlice only actually has one Buffer, this function has nothing to
+// do and simply returns said Buffer, hence it being "lazy".
+func (s BufferSlice) LazyMaterialize(pool BufferPool) *Buffer {
+	if len(s) == 1 {
+		return s[0].Ref()
+	}
+	buf := pool.Get(s.Len())
+	s.CopyTo(buf)
+	return NewBuffer(buf, pool.Put)
+}
+
+// Reader returns a new Reader for the input slice.
+func (s BufferSlice) Reader() *Reader {
+	return &Reader{
+		data: s,
+		len:  s.Len(),
+	}
+}
+
+var _ io.ReadCloser = (*Reader)(nil)
+
+// Reader exposes a BufferSlice's data as an io.Reader, allowing it to interface
+// with other parts systems. It also provides an additional convenience method
+// Remaining which returns the number of unread bytes remaining in the slice.
+//
+// Note that reading data from the reader does not free the underlying buffers!
+// Only calling Close once all data is read will free the buffers.
+type Reader struct {
+	data         BufferSlice
+	len          int
+	dataIdx, idx int
```

Comment:
One of these is a index into the slice of buffers, while the other is an index into a single byte slice. I feel that we need better names to clearly disambiguate these two. What do you think?